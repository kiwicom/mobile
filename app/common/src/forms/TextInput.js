// @flow

import * as React from 'react';
import { Icon } from '@kiwicom/react-native-app-common';
import { TextInput as OriginalTextInput, View } from 'react-native';

export const styles = {
  input: {
    color: 'black',
    backgroundColor: 'transparent',
    fontSize: 15,
    padding: 10,
    flex: 1,
  },
  wrapper: {
    height: 40,
    backgroundColor: '#fff',
    marginBottom: 5,
    flexDirection: 'row',
  },
  icon: {
    marginLeft: 10,
    alignSelf: 'center',
  },
};

// not exact - additional properties allowed
type Props = {
  iconName?: string, // only Material icons allowed here
};

export default function TextInput(props: Props) {
  return (
    <View style={styles.wrapper}>
      {props.iconName && (
        <Icon name={props.iconName} size={20} style={styles.icon} />
      )}
      <OriginalTextInput
        underlineColorAndroid="transparent"
        {...props}
        style={styles.input}
      />
    </View>
  );
}
