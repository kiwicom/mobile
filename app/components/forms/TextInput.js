// @flow

import * as React from 'react';
import { TextInput as OriginalTextInput, View } from 'react-native';

export const colours = {
  text: 'black',
  background: 'transparent',
};

export default function TextInput(props: Object) {
  return (
    <View
      style={{
        height: 40,
        backgroundColor: '#ededed',
        padding: 10,
        marginBottom: 5,
      }}
    >
      <OriginalTextInput
        underlineColorAndroid="transparent"
        {...props}
        style={{
          flex: 1,
          color: colours.text,
          backgroundColor: colours.background,
          fontSize: 15,
        }}
      />
    </View>
  );
}
