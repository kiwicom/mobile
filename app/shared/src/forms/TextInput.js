// @flow

import * as React from 'react';
import { Icon } from '@kiwicom/react-native-app-shared';
import { TextInput as OriginalTextInput, View } from 'react-native';

import StyleSheet from '../PlatformStyleSheet';

const styles = StyleSheet.create({
  input: {
    flex: 1,
    color: '#30363d',
    backgroundColor: 'transparent',
    padding: 10,
    android: {
      fontSize: 16,
    },
    ios: {
      fontSize: 14,
    },
  },
  wrapper: {
    flexDirection: 'row',
    backgroundColor: '#fff',
    elevation: 1, // Android only
    android: {
      borderRadius: 3,
      height: 48,
    },
    ios: {
      borderRadius: 0,
      height: 47,
    },
  },
  icon: {
    marginLeft: 10,
    alignSelf: 'center',
  },
});

// not exact - additional properties allowed
type Props = {
  iconName?: string, // only Material icons allowed here
};

export default function TextInput(props: Props) {
  return (
    <View style={styles.wrapper}>
      {props.iconName && (
        <Icon name={props.iconName} size={20} style={styles.icon} />
      )}
      <OriginalTextInput
        underlineColorAndroid="transparent"
        autoCorrect={false}
        {...props}
        style={styles.input}
      />
    </View>
  );
}
