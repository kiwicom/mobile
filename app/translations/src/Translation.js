// @flow

import * as React from 'react';
import { NativeModules } from 'react-native';
import { Text } from '@kiwicom/react-native-app-shared';

import DefaultVocabulary, { type TranslationKeys } from './DefaultVocabulary';
import CaseTransform, {
  type SupportedTransformationsType,
} from './transformations/CaseTransform';

type Props =
  | {|
      id: TranslationKeys,
      values?: Object,
      textTransform?: SupportedTransformationsType,
    |}
  | {|
      passThrough: ?string | ?number, // may be undefined because of IDX output
      textTransform?: SupportedTransformationsType,
    |};

/**
 * Regular translation using registered keys:
 *
 * ```js
 * <Translation id="SingleHotel.Rating.Stars" />
 * ```
 *
 * So called "pass through" translation (this props is not being translated
 * but passed through the component).
 *
 * ```js
 * <Translation passThrough="★★★★★" />
 * ```
 *
 * It's possible to apply text transformations in both cases:
 *
 * ```js
 * <Translation passThrough="★★★★★" textTransform="uppercase" />
 * <Translation id="SingleHotel.Rating.Stars" textTransform="uppercase" />
 * ```
 */
export default class Translation extends React.Component<Props> {
  // use this property to highlight translations for screenshoting or debugging
  highlightTranslations = false;

  replaceValues = (text: string, values?: Object): string => {
    const VARIABLE_REGEX = /__([0-9]+_)?([a-zA-Z-_]+)__/g;
    return text.replace(
      VARIABLE_REGEX,
      (match, variableIndex, variableName) => {
        // 'variableIndex' is currently being ignored (we don't need it)
        const value = values && values[variableName];
        return value !== undefined ? value : '?';
      },
    );
  };

  render = () => {
    const containerStyle = this.highlightTranslations
      ? {
          backgroundColor: this.props.id ? 'orange' : 'pink',
        }
      : undefined;

    if (this.props.id === undefined) {
      // ID prop is missing so let's assume it's "pass through" translation
      // (property passThrough may be undefined)
      return (
        <Text style={containerStyle}>
          {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
          {CaseTransform(
            String(this.props.passThrough),
            this.props.textTransform,
          )}
        </Text>
      );
    }

    const key = this.props.id;
    let translatedString = NativeModules.RNTranslationManager.translate(key);

    if (translatedString === key) {
      // fallback to our dummy vocabulary because native code didn't provide translation
      translatedString = DefaultVocabulary[key];
    }

    if (translatedString === undefined) {
      // Everything failed - native translation even our fallback. This usually
      // means we are using key that does not exist. In this case the only thing
      // we can actually do is to use original key as a text.
      translatedString = key;
    }

    const values = this.props.values || {};
    return (
      <Text style={containerStyle}>
        {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
        {CaseTransform(
          this.replaceValues(translatedString, values),
          this.props.textTransform,
        )}
      </Text>
    );
  };
}
