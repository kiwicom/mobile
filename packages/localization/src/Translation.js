// @flow

import * as React from 'react';
import { Text } from '@kiwicom/mobile-shared';
import { translate } from '@kiwicom/rnmodules';

import vocabulary, { type TranslationKeys } from './DefaultVocabulary';
import CaseTransform, {
  type SupportedTransformationsType,
} from './transformations/CaseTransform';
import { getTranslation, replaceValues } from './TranslationHelpers';

// Preload translations
Object.keys(vocabulary).forEach(async key => {
  const nativeKey = 'mobile.' + key;

  const translatedString = await translate(nativeKey);
  if (translatedString != null) {
    vocabulary[key] = getTranslation(translatedString, key);
  }
});

type CommonProps = {|
  testID?: string,
  textTransform?: SupportedTransformationsType,
|};

type Props =
  | {|
      id: TranslationKeys,
      values?: Object,
      ...CommonProps,
    |}
  | {|
      passThrough: ?string | ?number, // may be undefined because of IDX output
      ...CommonProps,
    |};

/**
 * Regular translation using registered keys:
 *
 * ```js
 * <Translation id="SingleHotel.Rating.Stars" />
 * ```
 *
 * So called "pass through" translation (this props is not being translated
 * but passed through the component).
 *
 * ```js
 * <Translation passThrough="★★★★★" />
 * ```
 *
 * It's possible to apply text transformations in both cases:
 *
 * ```js
 * <Translation passThrough="★★★★★" textTransform="uppercase" />
 * <Translation id="SingleHotel.Rating.Stars" textTransform="uppercase" />
 * ```
 */
export default class Translation extends React.Component<Props> {
  // use this property to highlight translations for screenshoting or debugging
  highlightTranslations = false;

  render = () => {
    const containerStyle = this.highlightTranslations
      ? {
          backgroundColor: this.props.id ? 'orange' : 'pink',
        }
      : undefined;

    if (this.props.id === undefined) {
      // ID prop is missing so let's assume it's "pass through" translation
      // (property passThrough may be undefined)
      return (
        <Text style={containerStyle}>
          {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
          {CaseTransform(
            String(this.props.passThrough),
            this.props.textTransform,
          )}
        </Text>
      );
    }

    const values = this.props.values || {};
    return (
      <Text style={containerStyle}>
        {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
        {CaseTransform(
          replaceValues(vocabulary[this.props.id], values),
          this.props.textTransform,
        )}
      </Text>
    );
  };
}
