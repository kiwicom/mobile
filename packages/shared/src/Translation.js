// @flow

import * as React from 'react';
import {
  CaseTransform,
  getTranslation,
  type SupportedTransformationsType,
  type TranslationKeysType,
} from '@kiwicom/mobile-localization';

import Text from './Text';

type CommonProps = {|
  testID?: string,
  textTransform?: SupportedTransformationsType,
|};

type Props =
  | {|
      id: TranslationKeysType,
      values?: Object,
      ...CommonProps,
    |}
  | {|
      passThrough: ?string | ?number, // may be undefined because of optional chaining
      ...CommonProps,
    |};

/**
 * Regular translation using registered keys:
 *
 * ```js
 * <Translation id="SingleHotel.Rating.Stars" />
 * ```
 *
 * So called "pass through" translation (this props is not being translated
 * but passed through the component).
 *
 * ```js
 * <Translation passThrough="★★★★★" />
 * ```
 *
 * It's possible to apply text transformations in both cases:
 *
 * ```js
 * <Translation passThrough="★★★★★" textTransform="uppercase" />
 * <Translation id="SingleHotel.Rating.Stars" textTransform="uppercase" />
 * ```
 */

// use this property to highlight translations for screenshoting or debugging
const highlightTranslations = false;
export default function Translation(props: Props) {
  const containerStyle = highlightTranslations
    ? {
        backgroundColor: props.id ? 'orange' : 'pink',
      }
    : undefined;
  if (props.id === undefined && props.passThrough != null) {
    return (
      <Text style={containerStyle}>
        {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
        {CaseTransform(String(props.passThrough), props.textTransform)}
      </Text>
    );
  } else if (props.id === undefined) {
    return null;
  }

  const values = props.values || {};
  return (
    <Text style={containerStyle}>
      {/* $FlowExpectedError: we do not allow to use 'string' in the 'Text' components but translations are exceptions. */}
      {CaseTransform(getTranslation(props.id, values), props.textTransform)}
    </Text>
  );
}
