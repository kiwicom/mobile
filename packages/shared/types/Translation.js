// @flow

import * as React from 'react';

import Translation from '../src/Translation';
import TranslationFragment from '../src/TranslationFragment';

export type TranslationType =
  | React.Element<typeof Translation>
  | React.Element<typeof TranslationFragment>;
