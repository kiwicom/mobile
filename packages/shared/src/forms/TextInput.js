// @flow strict

import * as React from 'react';
import { TextInput as OriginalTextInput, View } from 'react-native'; // eslint-disable-line no-restricted-imports
import { Translation } from '@kiwicom/mobile-localization';
import { defaultTokens } from '@kiwicom/mobile-orbit';

import { type StylePropType } from '../../types/Styles';
import StyleSheet from '../PlatformStyleSheet';
import Text from '../Text';

type Props = {|
  +placeholder?: React.Element<typeof Translation>,
  +defaultValue?: string,
  +autoFocus?: boolean,
  +onChangeText?: (text: string) => void,
  +keyboardType?: 'email-address' | 'numeric' | 'number-pad',
  +secureTextEntry?: boolean,
  +label?: React.Element<typeof Translation>,
  +labelStyle?: StylePropType,
  +inputWrapperStyle?: StylePropType,
|};

type State = {|
  displayPlaceholder: boolean,
|};

export default class TextInput extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    const { defaultValue } = props;
    this.state = {
      displayPlaceholder: defaultValue == null || defaultValue === '',
    };
  }

  handlePlaceholder = (text: string) => {
    this.setState(
      {
        displayPlaceholder: text === '',
      },
      () => {
        if (this.props.onChangeText) {
          this.props.onChangeText(text);
        }
      },
    );
  };

  render = () => (
    <React.Fragment>
      {this.props.label != null && (
        <Text style={[styleSheet.label, this.props.labelStyle]}>
          {this.props.label}
        </Text>
      )}
      <View style={[styleSheet.wrapper, this.props.inputWrapperStyle]}>
        <OriginalTextInput
          underlineColorAndroid="transparent"
          autoCorrect={false}
          {...this.props}
          placeholder={null}
          onChangeText={this.handlePlaceholder}
          style={[styleSheet.text, styleSheet.input]}
        />

        {this.props.placeholder &&
          this.state.displayPlaceholder && (
            <Text
              style={[styleSheet.text, styleSheet.placeholder]}
              pointerEvents="none"
            >
              {this.props.placeholder}
            </Text>
          )}
      </View>
    </React.Fragment>
  );
}

const styleSheet = StyleSheet.create({
  text: {
    android: {
      fontSize: 16,
    },
    ios: {
      fontSize: 14,
    },
  },
  input: {
    flex: 1,
    color: defaultTokens.colorTextAttention,
    backgroundColor: defaultTokens.paletteCloudNormal,
    borderRadius: 6,
    padding: 10,
  },
  wrapper: {
    flexDirection: 'row',
    android: {
      elevation: 1,
      height: 48,
    },
    ios: {
      height: 47,
    },
  },
  placeholder: {
    position: 'absolute',
    padding: 10,
    paddingVertical: 15,
    color: defaultTokens.colorTextSecondary,
  },
  label: { color: defaultTokens.colorTextSecondary },
});
