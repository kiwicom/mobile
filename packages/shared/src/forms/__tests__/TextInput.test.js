// @flow strict

import * as React from 'react';
import { PlaygroundRenderer } from '@kiwicom/mobile-playground';

import Translation from '../../Translation';
import TextInput from '../TextInput';

it('renders', () => {
  PlaygroundRenderer.render(<TextInput />);
});

it('renders with default value', () => {
  PlaygroundRenderer.render(<TextInput defaultValue="Default value" />);
});

it('renders with placeholder', () => {
  PlaygroundRenderer.render(
    <TextInput placeholder={<Translation passThrough="Placeholder" />} />,
  );
});

it('renders with label', () => {
  PlaygroundRenderer.render(
    <TextInput
      label={<Translation passThrough="label" />}
      defaultValue="Test"
    />,
  );
});
