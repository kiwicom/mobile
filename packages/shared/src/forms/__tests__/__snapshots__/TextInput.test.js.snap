// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<React.Fragment>
  <Component
    style={
      Object {
        "flexDirection": "row",
        "height": 47,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCorrect={false}
      onChangeText={[Function]}
      placeholder={null}
      style={
        Array [
          Object {
            "fontSize": 14,
          },
          Object {
            "backgroundColor": "#e8edf1",
            "borderRadius": 6,
            "color": "#171b1e",
            "flex": 1,
            "padding": 10,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </Component>
</React.Fragment>
`;

exports[`renders with default value 1`] = `
<React.Fragment>
  <Component
    style={
      Object {
        "flexDirection": "row",
        "height": 47,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCorrect={false}
      defaultValue="Default value"
      onChangeText={[Function]}
      placeholder={null}
      style={
        Array [
          Object {
            "fontSize": 14,
          },
          Object {
            "backgroundColor": "#e8edf1",
            "borderRadius": 6,
            "color": "#171b1e",
            "flex": 1,
            "padding": 10,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </Component>
</React.Fragment>
`;

exports[`renders with label 1`] = `
<React.Fragment>
  <Text
    style={
      Array [
        Object {
          "color": "#7f91a8",
        },
        undefined,
      ]
    }
  >
    <Translation
      passThrough="label"
    />
  </Text>
  <Component
    style={
      Object {
        "flexDirection": "row",
        "height": 47,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCorrect={false}
      defaultValue="Test"
      label={
        <Translation
          passThrough="label"
        />
      }
      onChangeText={[Function]}
      placeholder={null}
      style={
        Array [
          Object {
            "fontSize": 14,
          },
          Object {
            "backgroundColor": "#e8edf1",
            "borderRadius": 6,
            "color": "#171b1e",
            "flex": 1,
            "padding": 10,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </Component>
</React.Fragment>
`;

exports[`renders with placeholder 1`] = `
<React.Fragment>
  <Component
    style={
      Object {
        "flexDirection": "row",
        "height": 47,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCorrect={false}
      onChangeText={[Function]}
      placeholder={null}
      style={
        Array [
          Object {
            "fontSize": 14,
          },
          Object {
            "backgroundColor": "#e8edf1",
            "borderRadius": 6,
            "color": "#171b1e",
            "flex": 1,
            "padding": 10,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
    <Text
      pointerEvents="none"
      style={
        Array [
          Object {
            "fontSize": 14,
          },
          Object {
            "color": "#7f91a8",
            "padding": 10,
            "paddingVertical": 15,
            "position": "absolute",
          },
        ]
      }
    >
      <Translation
        passThrough="Placeholder"
      />
    </Text>
  </Component>
</React.Fragment>
`;
